src/Umbraco.Web/Editors/CodeFileController.cs

```diff
@@ -0,0 +1,155 @@
+﻿using AutoMapper;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Text;
+using System.Threading.Tasks;
+using System.Web.Http;
+using Umbraco.Core.Models;
+using Umbraco.Web.Models.ContentEditing;
+using Umbraco.Web.Mvc;
+using Umbraco.Web.WebApi;
+
+namespace Umbraco.Web.Editors
+{
+    [PluginController("UmbracoApi")]
+    public class CodeFileController : UmbracoAuthorizedJsonController
+    {
+
+        public HttpResponseMessage PostCreate(string type, CodeFileDisplay display)
+        {
+            switch (type)
+            {
+                case "partialView":
+                    var view = new PartialView(display.VirtualPath);
+                    var result = Services.FileService.CreatePartialView(view, display.Snippet, Security.CurrentUser.Id); 
+                    return result.Success == true ? Request.CreateResponse(HttpStatusCode.OK, result.Result) : Request.CreateNotificationValidationErrorResponse(result.Exception.Message);
+                case "partialViewMacro":
+                    var viewMacro = new PartialView(display.VirtualPath);
+                    var resultMacro = Services.FileService.CreatePartialViewMacro(viewMacro, display.Snippet, Security.CurrentUser.Id);
+                    return resultMacro.Success == true ? Request.CreateResponse(HttpStatusCode.OK, resultMacro.Result) : Request.CreateNotificationValidationErrorResponse(resultMacro.Exception.Message);
+                case "script":
+                    var script = new Script(display.VirtualPath);
+                    Services.FileService.SaveScript(script, Security.CurrentUser.Id);
+                    return Request.CreateResponse(HttpStatusCode.OK);
+                default:
+                    throw new ArgumentException("File Type not supported", "type");
```

Comment:
This should return a Not Found (404), we don't need to give malicious people more information than they need.